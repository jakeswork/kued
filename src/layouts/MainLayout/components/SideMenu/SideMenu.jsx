import React, { Component, Fragment } from 'react';
import { slide as Menu } from 'react-burger-menu';
import { FiX, FiMenu } from 'react-icons/fi';

import { propTypes, defaultProps } from './types';
import Viewport from '../../../../utils/Viewport';
import logo from '../../../../assets/kued-logotype.png';
import Text from '../../../../components/Text';
import MenuItems from './components/MenuItems';

class SideMenu extends Component {
  static propTypes = propTypes;

  static defaultProps = defaultProps;

  state = {
    isMobileViewport: false,
    isOpen: true,
  }

  componentDidMount() {
    this.updateViewportState();
    window.addEventListener('resize', this.updateViewportState);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.updateViewportState);
  }

  updateViewportState = () => {
    this.setState({
      isMobileViewport: Viewport.mobile(),
      isOpen: !Viewport.mobile(),
    });
  }

  render() {
    const { isOpen, isMobileViewport } = this.state;
    const { classes } = this.props;

    return (
      <Fragment>
        {isMobileViewport && (
          <FiMenu
            className={classes.menuIcon}
            data-test-id="openMenuButton"
            onClick={() => this.setState({ isOpen: !isOpen })}
          />
        )}
        <Menu
          left
          isOpen={isOpen}
          className={classes.burgerMenu}
          customBurgerIcon={false}
          customCrossIcon={false}
          noOverlay
          data-test-id="sideMenu"
          itemListClassName={classes.menuList}
        >
          {isMobileViewport && (
            <FiX
              className={classes.menuIcon}
              data-test-id="closeMenuButton"
              onClick={() => this.setState({ isOpen: false })}
              style={{ color: 'white' }}
            />
          )}
          <img src={logo} alt="Kued Logo" className={classes.logo} />
          <div className={classes.menuItems}>
            <MenuItems />
          </div>
          <div className={classes.sideMenuFooter}>
            <Text caption style={{ color: 'white' }}>v1.0.0</Text>
          </div>
        </Menu>
      </Fragment>
    );
  }
}

export default SideMenu;
