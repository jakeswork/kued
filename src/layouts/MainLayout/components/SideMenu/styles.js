export default theme => ({
  menuIcon: {
    fontSize: 24,
    position: 'fixed',
    top: 16,
    left: 16,
    cursor: 'pointer',
    color: theme.colorSecondary,
    zIndex: 2,
  },
  burgerMenu: {
    borderRight: `1px solid ${theme.textSecondary}`,
    background: theme.colorSecondary,
    display: 'flex',
    justifyContent: 'center',
    flexDirection: 'column',
    alignItems: 'center',
    position: 'relative',
    padding: '16px 0',
  },
  logo: {
    position: 'fixed',
    top: 16,
    outline: 'none',
    border: 0,
  },
  logoImg: {
    width: 160,
  },
  sideMenuFooter: {
    bottom: 16,
    position: 'fixed',
  },
  menuList: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  menuItems: {
    display: 'flex !important',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'column',
  },
});
