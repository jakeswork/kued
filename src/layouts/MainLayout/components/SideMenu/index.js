import withStyles from 'react-jss';

import SideMenu from './SideMenu';
import styles from './styles';

export default withStyles(styles)(SideMenu);
