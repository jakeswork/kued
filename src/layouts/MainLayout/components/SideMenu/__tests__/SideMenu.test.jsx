import React from 'react';
import { shallow } from 'enzyme';

import SideMenu from '../SideMenu';

const defaultProps = {
  classes: {},
};

describe('The SideMenu instance', () => {
  describe('When called with default properties', () => {
    it('should render without throwing errors', () => {
      const wrapper = shallow(<SideMenu {...defaultProps} />);

      expect(wrapper).toMatchSnapshot();

      wrapper.unmount();
    });
  });

  describe('when there is a mobile viewport', () => {
    describe('when you click the menu button', () => {
      it('should open the menu', () => {
        const wrapper = shallow(<SideMenu {...defaultProps} />);

        global.innerWidth = 414;

        global.dispatchEvent(new Event('resize'));

        const button = wrapper.find({ 'data-test-id': 'openMenuButton' });

        expect(wrapper.state('isOpen')).toBe(false);

        button.simulate('click');

        expect(wrapper.state('isOpen')).toBe(true);
      });
    });

    describe('when you click the close button', () => {
      it('should close the menu', () => {
        const wrapper = shallow(<SideMenu {...defaultProps} />);

        global.innerWidth = 414;

        global.dispatchEvent(new Event('resize'));

        const button = wrapper.find({ 'data-test-id': 'closeMenuButton' });

        wrapper.setState({ isOpen: true });

        expect(wrapper.state('isOpen')).toBe(true);

        button.simulate('click');

        expect(wrapper.state('isOpen')).toBe(false);
      });
    });
  });
});
