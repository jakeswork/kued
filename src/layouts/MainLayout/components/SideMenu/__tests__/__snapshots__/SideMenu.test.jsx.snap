// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SideMenu instance When called with default properties should render without throwing errors 1`] = `
<Fragment>
  <Menu
    bodyClassName=""
    burgerBarClassName=""
    burgerButtonClassName=""
    className=""
    crossButtonClassName=""
    crossClassName=""
    customBurgerIcon={false}
    customCrossIcon={false}
    data-test-id="sideMenu"
    disableAutoFocus={false}
    disableCloseOnEsc={false}
    htmlClassName=""
    id=""
    isOpen={true}
    itemClassName=""
    itemListClassName=""
    left={true}
    menuClassName=""
    morphShapeClassName=""
    noOverlay={true}
    onStateChange={[Function]}
    outerContainerId=""
    overlayClassName=""
    pageWrapId=""
    styles={Object {}}
    width={300}
  >
    <Link
      to="/"
    >
      <img
        alt="Kued Logo"
        src="kued-logotype.png"
      />
    </Link>
    <div>
      <Jss(MenuItems) />
    </div>
    <div>
      <Jss(Text)
        bold={false}
        caption={true}
        h1={false}
        h2={false}
        h3={false}
        h4={false}
        h5={false}
        style={
          Object {
            "color": "white",
          }
        }
      >
        v1.0.0
      </Jss(Text)>
    </div>
  </Menu>
</Fragment>
`;
